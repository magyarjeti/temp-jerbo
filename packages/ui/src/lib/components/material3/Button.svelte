<script lang="ts">
	import type { Snippet } from 'svelte';
	import JetiIcon from '../JetiIcon.svelte';
	import StateLayer from './StateLayer.svelte';
	import { concatClasses } from '@packages/toolbox/concatClasses';
	import type { HTMLButtonAttributes, HTMLAnchorAttributes } from 'svelte/elements';
	import { isAnchorAttrs, isButtonAttrs } from '../../utils/components';

	type Props = {
		variant?: 'filled' | 'outlined' | 'elevated' | 'text';
		disabled?: boolean | null;
		icon?: string | null | Snippet;
		selected?: boolean;
		children?: Snippet;
	} & (HTMLButtonAttributes | HTMLAnchorAttributes);

	let {
		variant = 'outlined',
		disabled,
		icon,
		selected = false,
		children,
		...attrs
	}: Props = $props();

	const variantClassMap: Record<typeof variant, string> = {
		filled: concatClasses(
			'bg-primary text-on-primary hover:elevation-1',
			selected && ' elevation-1',
		),
		elevated: concatClasses(
			'bg-surface-container text-primary elevation-1 hover:elevation-2',
			selected && ' elevation-2',
		),
		outlined: concatClasses(
			'border-outline-variant border bg-transparent focus:border-primary',
			selected && ' border-primary',
		),
		text: concatClasses(
			'text-primary',
			icon ? 'pl-3 pr-4' : 'px-3',
			selected && ' bg-primary text-on-primary',
		),
	};
	const stateLayerClasses: Record<typeof variant, string> = {
		filled: 'bg-on-primary',
		elevated: 'bg-primary',
		outlined: 'bg-primary',
		text: 'bg-primary',
	};
	const baseClasses = concatClasses(
		'relative typo-label-large shape-full min-h-10 gap-2 py-1 inline-flex cursor-pointer max-w-full items-center justify-center disabled:text-on-surface/38',
	);
	const nonTextVariantBaseClasses = concatClasses(
		'disabled:bg-on-surface/12',
		icon ? 'pl-4 pr-6' : 'px-6',
	);
	const classes = $derived(
		concatClasses(
			attrs.class,
			baseClasses,
			variantClassMap[variant],
			variant !== 'text' && nonTextVariantBaseClasses,
		),
	);
</script>

{#snippet inside()}
	<StateLayer class="{stateLayerClasses[variant]} shape-full" />
	{#if typeof icon === 'string'}
		<JetiIcon>{icon}</JetiIcon>
	{:else if icon}
		{@render icon()}
	{/if}
	<span class="min-w-0 {'href' in attrs ? 'truncate' : 'block'}">{@render children?.()}</span>
{/snippet}

{#if isAnchorAttrs(attrs)}
	<a
		aria-disabled={disabled}
		{...attrs}
		class={concatClasses(
			classes,
			disabled && 'pointer-events-none cursor-not-allowed text-on-surface/38',
			disabled && variant !== 'text' && 'bg-on-surface/12',
		)}
	>
		{@render inside()}
	</a>
{:else if isButtonAttrs(attrs)}
	<button
		aria-disabled={disabled}
		type={attrs.type || 'button'}
		{disabled}
		{...attrs}
		class={classes}
	>
		{@render inside()}
	</button>
{/if}
