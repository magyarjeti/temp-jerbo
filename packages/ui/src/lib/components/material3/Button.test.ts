import { render } from '@testing-library/svelte';
import { expect, test } from 'vitest';
import Button from './Button.svelte';

test('Elevated Button snapshot', () => {
	const { container } = render(Button, { props: { variant: 'elevated' } });
	expect(container).toMatchSnapshot();
});

test('Outlined Button snapshot', () => {
	const { container } = render(Button, { props: { variant: 'outlined' } });
	expect(container).toMatchSnapshot();
});

test('Filled Button snapshot', () => {
	const { container } = render(Button, { props: { variant: 'filled' } });
	expect(container).toMatchSnapshot();
});

test('Text Button snapshot', () => {
	const { container } = render(Button, { props: { variant: 'text' } });
	expect(container).toMatchSnapshot();
});
