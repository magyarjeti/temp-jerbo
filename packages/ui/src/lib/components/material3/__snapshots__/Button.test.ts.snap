// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Elevated Button snapshot 1`] = `
<div>
  <button
    class="max-w-full typo-label-large inline-flex min-h-10 cursor-pointer items-center justify-center gap-2 py-1 shape-full bg-surface-container text-primary state-layers state-layer-focus-primary state-layer-hover-primary elevation-1 hover:elevation-2 px-6"
  >
     
    <span
      class="block min-w-0"
    />
  </button>
  
</div>
`;

exports[`Filled Button snapshot 1`] = `
<div>
  <button
    class="max-w-full typo-label-large inline-flex min-h-10 cursor-pointer items-center justify-center gap-2 py-1 shape-full bg-primary text-on-primary state-layers state-layer-focus-on-primary state-layer-hover-on-primary hover:elevation-1 disabled:pointer-events-none disabled:bg-on-surface/12 disabled:text-on-surface/38 px-6"
  >
     
    <span
      class="block min-w-0"
    />
  </button>
  
</div>
`;

exports[`Outlined Button snapshot 1`] = `
<div>
  <button
    class="max-w-full typo-label-large inline-flex min-h-10 cursor-pointer items-center justify-center gap-2 py-1 shape-full bg-transparent border border-outline-variant text-primary focus:border-primary state-layers state-layer-hover-primary state-layer-focus-primary px-6"
  >
     
    <span
      class="block min-w-0"
    />
  </button>
  
</div>
`;

exports[`Text Button snapshot 1`] = `
<div>
  <button
    class="max-w-full typo-label-large inline-flex min-h-10 cursor-pointer items-center justify-center gap-2 py-1 text-primary shape-full state-layers state-layer-focus-primary state-layer-hover-primary disabled:text-on-surface/38 px-3"
  >
     
    <span
      class="block min-w-0"
    />
  </button>
  
</div>
`;
